script/install/debootstrap.bash

```diff
@@ -171,5 +171,47 @@ echo "$chroot_cmd update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-9
 chroot "$ISOLATE_ROOT" update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-9 75
 # gcc 9 done
 
-# let user know that chroot installs are finished
-
+# JavaScript V8 engine compile and install
+echo "Preparing to compile and install the JavaScript V8 engine."
+echo
+echo "If this fails, visit https://v8.dev/docs/build and follow the instructions."
+echo "This may take a while ..."
```

Comment:
Good point, it actually hits the job time limit on Travis.
https://travis-ci.org/NZOI/nztrain/jobs/640430838

Reply:
Since it takes so long I think it's pretty important that we also have a way to disable V8 in normal installations. I'm not sure what's the best way to do this (we would also need to remove JavaScript from the language dropdown).

Reply:
Travis set a [`$CI` environment variable](https://docs.travis-ci.com/user/environment-variables/#default-environment-variables) which we can branch on. I think that can take us half-way.

Reply:
I've tested it on an 8-core VM and it the V8 build took about 15 minutes, does that still seem too long?

> we would also need to remove JavaScript from the language dropdown

If we're only disabling V8 on development installs I think it's fine to leave it (maybe with a message that "submissions in JavaScript V8 will not work").

Reply:
I tried re-installing [from this ](https://github.com/NZOI/nztrain/commits/add-javascript-into-master) on a fresh Ubuntu 16.04 install and the whole process took 46 minutes.
- openjdk-11 took 20 minutes to download from ppa.launchpad.net (at ~230 kB/s).
- V8 took 7 minutes to download and another 7 minutes to build (using 8 cores).

If we want to add options to not install those I think that could be part of a separate pull request? In either case it should be pretty easy to add if we use [`confirm.bash`](https://github.com/NZOI/nztrain/blob/master/script/confirm.bash).

Reply:
I agree this should be part of a separate pull request. A longer term solution could be to only install GCC (or another language) by default. A separate install script can optionally be involved for all other languages - which means any future languages we add will also be left out of travis by default. The extra install step would need to be added to the README

Reply:
For now, I can confirm that the following change works.
```diff
diff --git a/script/install/debootstrap.bash b/script/install/debootstrap.bash
index c07dba5..6fe92b9 100644
--- a/script/install/debootstrap.bash
+++ b/script/install/debootstrap.bash
@@ -103,8 +103,10 @@ if ! chroot "$ISOLATE_ROOT" apt-cache show openjdk-11-jdk &>/dev/null; then
   chroot "$ISOLATE_ROOT" apt-get update
 fi
 
-echo "$chroot_install openjdk-11-jdk"
-chroot "$ISOLATE_ROOT" apt-get install openjdk-11-jdk # Java
+bash script/confirm.bash "Install Java (submissions in Java will fail without this!)" && {
+  echo "$chroot_install openjdk-11-jdk"
+  chroot "$ISOLATE_ROOT" apt-get install openjdk-11-jdk # Java
+}
 
 [ -z "$TRAVIS" ] && { # if not in Travis-CI
 
@@ -179,7 +181,7 @@ echo "$chroot_cmd update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-9
 chroot "$ISOLATE_ROOT" update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-9 75
 # gcc 9 done
 
-[ -z "$TRAVIS" ] && { # if not in Travis-CI
+[ -z "$TRAVIS" ] && bash script/confirm.bash "Install JavaScript V8 (submissions in V8 will fail without this!)" && {
   # JavaScript V8 engine compile and install
   echo "Preparing to compile and install the JavaScript V8 engine."
   echo
```

Since the changes are small I don't mind adding it here.

Reply:
> I've tested it on an 8-core VM and the V8 build took about 15 minutes, does that still seem too long?

For me it takes one hour on a single-core VM. Using `confirm.bash` will do, I don't mind if it's in this pull request or a separate one.

> > we would also need to remove JavaScript from the language dropdown
>
> If we're only disabling V8 on development installs I think it's fine to leave it (maybe with a message that "submissions in JavaScript V8 will not work").

Fair enough. If people want to deploy without JavaScript they can manually edit app/model/language.rb.

> openjdk-11 took 20 minutes to download from ppa.launchpad.net (at ~230 kB/s).

We should try to improve that. I'll create a separate pull request.